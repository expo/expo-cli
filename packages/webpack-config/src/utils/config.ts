export const DEFAULT_ALIAS = {
  // Alias direct react-native imports to react-native-web
  'react-native$': 'react-native-web',
  '@react-native-community/netinfo': 'react-native-web/dist/exports/NetInfo',
  // Add polyfills for modules that react-native-web doesn't support
  // Depends on expo-asset
  'react-native/Libraries/Image/AssetSourceResolver$': 'expo-asset/build/AssetSourceResolver',
  'react-native/Libraries/Image/assetPathUtils$': 'expo-asset/build/Image/assetPathUtils',
  'react-native/Libraries/Image/resolveAssetSource$': 'expo-asset/build/resolveAssetSource',
  // Alias internal react-native modules to react-native-web
  'react-native/Libraries/Components/View/ViewStylePropTypes$':
    'react-native-web/dist/exports/View/ViewStylePropTypes',
  'react-native/Libraries/EventEmitter/RCTDeviceEventEmitter$':
    'react-native-web/dist/vendor/react-native/NativeEventEmitter/RCTDeviceEventEmitter',
  'react-native/Libraries/vendor/emitter/EventEmitter$':
    'react-native-web/dist/vendor/react-native/emitter/EventEmitter',
  'react-native/Libraries/vendor/emitter/EventSubscriptionVendor$':
    'react-native-web/dist/vendor/react-native/emitter/EventSubscriptionVendor',
  'react-native/Libraries/EventEmitter/NativeEventEmitter$':
    'react-native-web/dist/vendor/react-native/NativeEventEmitter',
};

export function getModuleFileExtensions(...platforms: string[]): string[] {
  let fileExtensions = [];

  // Support both TypeScript and JavaScript
  for (const extension of ['ts', 'tsx', 'js', 'jsx']) {
    // Ensure order is correct: [platformA.js, platformB.js, js]
    for (const platform of [...platforms, '']) {
      fileExtensions.push([platform, extension].filter(Boolean).join('.'));
    }
  }
  // Always add this last
  fileExtensions.push('json');
  // Webpack requires a `.` before each value
  return fileExtensions.map(value => `.${value}`);
}

function isObject(val: any): boolean {
  if (val === null) {
    return false;
  }
  return typeof val === 'function' || typeof val === 'object';
}

/**
 * Given a config option that could evalutate to true, config, or null; return a config.
 * e.g.
 * `polyfill: true` returns the `config`
 * `polyfill: {}` returns `{}`
 */

export function enableWithPropertyOrConfig(
  prop: any,
  config: boolean | { [key: string]: any },
  merge: boolean = false
): any {
  // Value is truthy but not a replacement config.
  if (prop) {
    if (isObject(prop) && merge) {
      if (config == null || typeof config !== 'object') {
        throw new Error('enableWithPropertyOrConfig cannot merge config: ' + config);
      }
      return {
        ...config,
        ...prop,
      };
    }
    // Return the default config
    return config;
  }
  // Return falsey or replacement config.
  return prop;
}

/**
 * Used for features that are enabled by default unless specified otherwise.
 */
export function overrideWithPropertyOrConfig(
  prop: any,
  config: boolean | { [key: string]: any },
  merge: boolean = false
): any {
  if (prop === undefined) {
    return config;
  }
  return enableWithPropertyOrConfig(prop, config, merge);
}
