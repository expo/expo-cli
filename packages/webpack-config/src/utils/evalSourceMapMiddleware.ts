/**
 * Copyright (c) 2022 Expo, Inc.
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Based on https://github.com/facebook/create-react-app/blob/a422bf2/packages/react-dev-utils/evalSourceMapMiddleware.js
 * But with Node LTS support.
 */

import type express from 'express';
import type webpack from 'webpack';
import type WebpackDevServer from 'webpack-dev-server';

function base64SourceMap(source: any) {
  const base64 = Buffer.from(JSON.stringify(source.map()), 'utf8').toString('base64');
  return `data:application/json;charset=utf-8;base64,${base64}`;
}

function getSourceById(server: WebpackDevServer & any, id: string) {
  const module = Array.from(server._stats.compilation.modules).find(
    m => server._stats.compilation.chunkGraph.getModuleId(m) === id
  ) as webpack.Module[];

  // @ts-ignore
  return module.originalSource();
}

/*
 * Middleware responsible for retrieving a generated source
 * Receives a webpack internal url: "webpack-internal:///<module-id>"
 * Returns a generated source: "<source-text><sourceMappingURL><sourceURL>"
 *
 * Based on EvalSourceMapDevToolModuleTemplatePlugin.js
 */
export function createEvalSourceMapMiddleware(server: WebpackDevServer) {
  return function handleWebpackInternalMiddleware(
    req: express.Request,
    res: express.Response,
    next: express.NextFunction
  ) {
    if (req.url.startsWith('/__get-internal-source')) {
      const fileName = req.query.fileName;
      if (typeof fileName !== 'string') {
        return next();
      }
      const id = fileName?.match(/webpack-internal:\/\/\/(.+)/)?.[1];
      // @ts-ignore: untyped
      if (!id || !server._stats) {
        return next();
      }

      const source = getSourceById(server, id);
      const sourceMapURL = `//# sourceMappingURL=${base64SourceMap(source)}`;
      const sourceURL = `//# sourceURL=webpack-internal:///${module.id}`;
      res.end(`${source.source()}\n${sourceMapURL}\n${sourceURL}`);
    } else {
      return next();
    }
  };
}
