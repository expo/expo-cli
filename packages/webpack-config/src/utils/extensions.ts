export function getModuleFileExtensionsWithoutDotPrefix(...platforms: string[]): string[] {
  let fileExtensions = [];

  // Support both TypeScript and JavaScript
  for (const extension of ['ts', 'tsx', 'js', 'jsx']) {
    // Ensure order is correct: [platformA.js, platformB.js, js]
    for (const platform of [...platforms, '']) {
      fileExtensions.push([platform, extension].filter(Boolean).join('.'));
    }
  }
  // Always add this last
  fileExtensions.push('json');

  return fileExtensions;
}

export function getModuleFileExtensions(...platforms: string[]): string[] {
  // Webpack requires a `.` before each value
  return getModuleFileExtensionsWithoutDotPrefix(...platforms).map(value => `.${value}`);
}
