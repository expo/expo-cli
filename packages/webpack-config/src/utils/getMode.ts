import { Mode } from '../types';

/**
 * mode -> production -> development -> process.env.NODE_ENV -> 'development'
 */

function getMode({
  production,
  development,
  mode,
}: {
  production?: boolean;
  development?: boolean;
  mode?: Mode;
}): Mode {
  if (mode === undefined) {
    if (process.env.NODE_ENV != null && isValidMode(process.env.NODE_ENV)) {
      return process.env.NODE_ENV.toLowerCase() as Mode;
    }
  } else if (isValidMode(mode)) {
    return mode.toLowerCase() as Mode;
  }
  if (production) {
    return 'production';
  } else if (development) {
    return 'development';
  }
  return 'development';
}

function isValidMode(inputMode?: string): boolean {
  let mode;
  if (typeof inputMode === 'string') {
    mode = inputMode.toLowerCase();
  }
  return mode === 'production' || mode === 'development';
}

export default getMode;
