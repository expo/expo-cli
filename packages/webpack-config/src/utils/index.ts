export * from './search';
export { default as normalizePaths } from './normalizePaths';
export * from './config';
