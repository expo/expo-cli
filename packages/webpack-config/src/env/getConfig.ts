import { ExpoConfig, getConfigForPWA } from '@expo/config';

import { Environment } from '../types';
import { getPaths } from './paths';

/**
 * Get the Expo project config in a way that's optimized for web.
 *
 * @param env Environment properties used for getting the Expo project config.
 * @category env
 */
function getConfig(env: Pick<Environment, 'projectRoot' | 'config' | 'locations'>): ExpoConfig {
  if (env.config) {
    return env.config;
  }
  const locations = env.locations || getPaths(env.projectRoot);
  // Fill all config values with PWA defaults
  return getConfigForPWA(env.projectRoot, locations.absolute, {
    templateIcon: locations.template.get('icon.png'),
  });
}

export default getConfig;
