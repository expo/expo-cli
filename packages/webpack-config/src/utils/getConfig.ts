import { ExpoConfig, getConfigForPWA } from '@expo/config';

import { Environment } from '../types';
import { getPaths } from './paths';

function getConfig(env: Environment): ExpoConfig {
  if (env.config) {
    return env.config;
  }
  const locations = env.locations || getPaths(env.projectRoot);
  // Fill all config values with PWA defaults
  return getConfigForPWA(env.projectRoot, locations.absolute, {
    templateIcon: locations.template.get('icon.png'),
  });
}

export default getConfig;
