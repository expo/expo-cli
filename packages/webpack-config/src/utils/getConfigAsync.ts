import { ExpoConfig, getConfigForPWA } from '@expo/config';

import { Environment } from '../types';
import getPathsAsync from './getPathsAsync';

async function getConfigAsync(env: Environment): Promise<ExpoConfig> {
  if (env.config) {
    return env.config;
  }
  const locations = await getPathsAsync(env);
  // Fill all config values with PWA defaults
  return getConfigForPWA(env.projectRoot, locations.absolute, {
    templateIcon: locations.template.get('icon.png'),
  });
}

export default getConfigAsync;
