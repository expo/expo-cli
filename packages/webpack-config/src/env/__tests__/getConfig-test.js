/* eslint-env node */

import path from 'path';
import getConfig from '../getConfig';
import { normalizePaths } from '../../utils';

const projectRoot = path.resolve(__dirname, '../../../tests/basic');
const mode = 'development';
const env = { projectRoot, mode };

it(`has consistent defaults`, () => {
  const config = getConfig(env);
  const normalized = normalizePaths(config, value => value.split('packages/webpack-config/').pop());
  expect(normalized).toMatchSnapshot();
});
