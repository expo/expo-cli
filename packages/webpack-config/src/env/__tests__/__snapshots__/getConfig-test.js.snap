// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has consistent defaults 1`] = `
Object {
  "android": Object {},
  "description": "A Neat Expo App",
  "icon": "https://github.com/expo/expo/blob/master/templates/expo-template-blank/assets/icon.png?raw=true",
  "ios": Object {},
  "name": "basic",
  "platforms": Array [
    "ios",
    "android",
    "web",
  ],
  "primaryColor": undefined,
  "sdkVersion": "36.0.0",
  "slug": "basic",
  "splash": Object {
    "backgroundColor": "#ffffff",
    "image": "https://github.com/expo/expo/blob/master/templates/expo-template-blank/assets/splash.png?raw=true",
    "resizeMode": "contain",
  },
  "version": "1.0.0",
  "web": Object {
    "backgroundColor": "#ffffff",
    "barStyle": "black-translucent",
    "build": Object {
      "output": "web-build",
      "publicPath": "/",
      "rootId": "root",
    },
    "crossorigin": undefined,
    "dangerous": Object {
      "noJavaScriptMessage": "Oh no! It looks like JavaScript is not enabled in your browser.",
    },
    "description": "A Neat Expo App",
    "dir": "auto",
    "display": "standalone",
    "icons": Array [
      Object {
        "destination": "apple/icons",
        "size": Array [
          192,
          512,
        ],
        "src": "https://github.com/expo/expo/blob/master/templates/expo-template-blank/assets/icon.png?raw=true",
      },
      Object {
        "destination": "apple/icons",
        "ios": true,
        "sizes": 180,
        "src": "https://github.com/expo/expo/blob/master/templates/expo-template-blank/assets/icon.png?raw=true",
      },
    ],
    "lang": "en",
    "meta": Object {
      "apple": Object {
        "barStyle": "black-translucent",
        "formatDetection": "telephone=no",
        "mobileWebAppCapable": "yes",
        "touchFullscreen": "yes",
      },
      "viewport": "width=device-width,initial-scale=1,minimum-scale=1,maximum-scale=1.00001,viewport-fit=cover",
    },
    "name": "basic",
    "orientation": "any",
    "preferRelatedApplications": true,
    "relatedApplications": Array [],
    "scope": undefined,
    "shortName": "basic",
    "startUrl": ".",
    "startupImages": Array [
      Object {
        "color": "#ffffff",
        "destination": "apple/splash",
        "orientation": "any",
        "resizeMode": "contain",
        "src": "https://github.com/expo/expo/blob/master/templates/expo-template-blank/assets/splash.png?raw=true",
        "supportsTablet": true,
      },
    ],
    "themeColor": undefined,
  },
}
`;
